### ISSUE TITLE
ubuntu 20.04 running hello_triangle fails with "Failed to find a create function for"

### ISSUE THREAD

**Issue by memory-thrasher** [KhronosGroup/Vulkan-Samples#289]
I built the samples project per the linux build guide. I have vulkan-sdk from the lunarg repo. Build did not show any errors.

```
$ ./build/linux/app/bin/Release/x86_64/vulkan_samples hello_triangle
[info] Logger initialized
[info] Window created
[error] [build/linux/app/main.cpp:73] Failed to find a create function for
```
(Yes, the last line ends with "for")

* [My vulkaninfo](https://pastebin.com/PNkhynqD)
* I have both clang and gcc installed. I'm not sure which one cmake picked.
* Xubuntu 20.04
* nvidia 1080 using proprietary driver 460.80
* radeon x570 using mesa (I think)

**Comment by jamesdolan** [comment]
FWIW, issue is due to there being code with side-effect inside `assert()` statements, here is a little one-line fix for whats causing your issue...
```
index d41ad7b..d97fb27 100644
--- a/framework/platform/parser.h
+++ b/framework/platform/parser.h
@@ -217,7 +217,8 @@ class CommandParser
                auto values = get_command_value(command);
 
                Type type{};
-               assert(convert_type(values, &type) && "Failed to retrieve value. Type unsupported");
+               auto status = convert_type(values, &type);
+               assert(status && "Failed to retrieve value. Type unsupported");
                return type;
        }
 
@@ -301,4 +302,4 @@ inline bool CommandParser::convert_type(const std::vector<std::string> &values,
        return true;
 }
```

**Comment by memory-thrasher** [comment]
Thanks @jamesdolan . That got it past that error, so I consider the above patch to be a fix, and suggest that the repo maintainers include it. I am on the next error now.

```
$ ./build/linux/app/bin/Release/x86_64/vulkan_samples hello_triangle
[info] Logger initialized
[info] Window created
[info] Initializing vulkan instance.
[info] Enabled Validation Layers:
[info] Found GPU: AMD RADV POLARIS10 (ACO)
[info] Found GPU: GeForce GTX 1080
[info] Initializing vulkan device.
[error] [samples/api/hello_triangle/hello_triangle.cpp:319] Did not find suitable queue which supports graphics, compute and presentation.
Segmentation fault (core dumped)
```

At least this error means something.

Edit: Oh, so the sample just assumes gpu 0 is the right one. hello_triangle.cpp:282
```
	// Select the first GPU we find in the system.
	context.gpu = gpus.front();
```
That's a separate issue anyway.

**Comment by memory-thrasher** [comment]
```diff --git a/framework/platform/parser.h b/framework/platform/parser.h
index d41ad7b..d97fb27 100644
--- a/framework/platform/parser.h
+++ b/framework/platform/parser.h
@@ -217,7 +217,8 @@ class CommandParser
                auto values = get_command_value(command);
 
                Type type{};
-               assert(convert_type(values, &type) && "Failed to retrieve value. Type unsupported");
+               auto status = convert_type(values, &type);
+               assert(status && "Failed to retrieve value. Type unsupported");
                return type;
        }
 
diff --git a/samples/api/hello_triangle/hello_triangle.cpp b/samples/api/hello_triangle/hello_triangle.cpp
index 3f2b8e7..78ea885 100644
--- a/samples/api/hello_triangle/hello_triangle.cpp
+++ b/samples/api/hello_triangle/hello_triangle.cpp
@@ -277,8 +277,9 @@ void HelloTriangle::init_device(Context &                        context,
        std::vector<VkPhysicalDevice> gpus(gpu_count);
        VK_CHECK(vkEnumeratePhysicalDevices(context.instance, &gpu_count, gpus.data()));
 
-       // Select the first GPU we find in the system.
-       context.gpu = gpus.front();
+       for(size_t i = 0;i < gpu_count && (context.graphics_queue_index < 0);i++) {
+
+         context.gpu = gpus[i];
 
          uint32_t queue_family_count;
          vkGetPhysicalDeviceQueueFamilyProperties(context.gpu, &queue_family_count, nullptr);
@@ -313,6 +314,7 @@ void HelloTriangle::init_device(Context &                        context,
                  break;
                }
            }
+       }
 
        if (context.graphics_queue_index < 0)
```

**Comment by SaschaWillems** [comment]
This will be fixed with PR #288

**Comment by gpx1000** [comment]
There's two issues reported here.  The first is indeed addressed in PR #288 however, the second, fixing the assumption that the first GPU found in the system is the correct one to use is indeed another issue that needs to be fixed by a PR.  I'd like to suggest submitting a PR if you're willing @memory-thrasher to ensure you maintain proper recognition and attribution for both your issue discovery and fix submission.

**Comment by memory-thrasher** [comment]
@gpx1000 
I am. I want to look into disposing of the unused gpu objects correctly, and then apply it to all the samples. Should be ready before the weekend.
